src/main.ts

```diff
@@ -1,37 +1,58 @@
 import * as core from "@actions/core";
 import { context } from "@actions/github";
-import { checkIfEcrImageExists, getEcrRegistry } from "./aws-utils";
+import { getEcrRegistry } from "./aws-utils";
 import { format } from "date-fns";
 import { Docker } from "docker-cli-js";
 import { actionInput } from "./action-input";
+import {buildDocker, DockerBuild, publishDocker} from './docker-utils'
 
 async function run(): Promise<void> {
-	const registry: string = await getEcrRegistry();
-	const dockerfile: string = actionInput.dockerfile;
+  const registry: string = await getEcrRegistry();
+  const imageTag: string = createImageTag();
 
-	const repository: string = actionInput.repository;
-	const imageTag: string = createImageTag();
-	await checkIfEcrImageExists(repository, imageTag);
+  const docker: Docker = new Docker();
+  const builds: DockerBuild[] = actionInput.targets.map((targetInput) => {
+    return {
+      imageTag: imageTag,
+      ...targetInput
+    };
+  });
 
-	// Build and push image
-	const regRepoImg = `${registry}/${repository}:${imageTag}`;
-	const dockerBuild = `build ${actionInput.buildArgs} -t ${regRepoImg} -f ./${dockerfile} .`;
+  // Build all images before publishing, to reduce likelihood of partial success
+  for (const build of builds) {
+    await buildDocker(docker, registry, imageTag, build);
+  }
 
-	const docker = new Docker();
-	await docker.command(dockerBuild);
-	await docker.command(`push ${regRepoImg}`);
+  // If ay of the images fail to publish, the run will be marked as failed,
+  // but the remaining images will still be uploaded.
+  // This will allow the action to recover from a previous failure by uploading
+  // the remaining images, even in cases where image tags are immutable (which
+  // will make images that have succeeded in previous runs to fail).
+  let hasErrors = false;
+  for (const build of builds) {
+    try {
+      await publishDocker(docker, registry, build);
+    } catch (e: any) {
+      core.error(e.message);
+      hasErrors = true;
+    }
+  }
```

Comment:
Maybe some success logging would be nice as well, as each build completes? Or would that be visible in the workflow itself, either by the process going forward, or docker logging when building?

Reply:
> Do you want these to be sequential like this

Yes, actually parallelizing the builds would probably slow it down. When building the second target, we want the first target to be completed so that the output of the earlier stages can be reused in subsequent builds. From what I can tell, buildkit will only look for cached layers, it doesn't check if a layer is in the process of being created.

> Success logging

Experimentation in progress. Basically the problem is that these produce something like 30k lines of logs, lol. But there seems to be some kind of grouping. Testing testing.

Reply:
Reverting experiment. This SHOULD work, but it ain't doin' shit.

```ts
  core.startGroup("Building images")
  for (const build of builds) {
    const index = builds.indexOf(build)
    core.startGroup(`Building image ${index + 1} of ${builds.length}`)
    await buildDocker(docker, registry, imageTag, build);
    core.endGroup()
  }
  core.endGroup()
```